src/queue.c

```diff
@@ -913,6 +913,52 @@ _dispatch_root_queues_init(void *context DISPATCH_UNUSED)
 
 #define countof(x) (sizeof(x) / sizeof(x[0]))
 
+#if DISPATCH_USE_THREAD_LOCAL_STORAGE
+#include <unistd.h>
+#include <sys/syscall.h>
+
+#ifdef SYS_gettid
+static inline pid_t gettid() { return (pid_t) syscall(SYS_gettid); }
+#else
+#error "SYS_gettid unavailable on this system"
+#endif
+
+#define _tsd_call_cleanup(k,f)  \
+	do { if (f && __dispatch_tsd.k) ((void(*)(void*))(f))(__dispatch_tsd.k); \
+    } while (0)
```

Comment:
to handle calls like this:
    _tsd_call_cleanup(dispatch_io_key, NULL);
in the _libdispatch_thread_cleanup.
One can argue I don't need that anyway, since it will be optimized out (see macro), but
I wanted it there to get some symmetry in the code for TSD and non-TSD code.
If you prefer I don't call the NULL instances of these I can simplify the macro.
Also look at the "dispatch_sema4_key  K,V" .. same ugliness there in the code by default"